src/linker/runner.py

```diff
@@ -37,15 +48,8 @@ def main(
             "only computing_environment 'local' and 'slurm' are supported; "
             f"provided {config.computing_environment}"
         )
-    for step_name in config.steps:
-        runner(
-            container_engine=config.container_engine,
-            input_data=config.input_data,
-            results_dir=results_dir,
-            step_name=step_name,
-            implementation_dir=config.get_implementation_directory(step_name),
-            container_full_stem=config.get_container_full_stem(step_name),
-        )
+
+    pipeline.run(runner, results_dir)
```

Comment:
I think it might be cleaner to save the `runner` as an attribute of pipeline - i.e:
```python
if config.computing_environment == "local":
    runner = run_container
else:
    ...
    runner = partial(launch_slurm_job, session, resources)

pipeline = Pipeline(config, runner)
pipeline.run(results_dir)
```

Or if we want to create the pipeline object up front, so any validation failures will happen before we interact with drmaa, we can do this:

```python
pipeline = Pipeline(config)
...

if config.computing_environment == "local":
    pipeline.set_runner(run_container)
else:
    ...
    pipeline.set_runner(partial(launch_slurm_job, session, resources))

pipeline.run(results_dir)
```

Reply:
Hmm, I did want the Pipeline instantiation to be up front but adding a `set_runner` method feels icky. But maybe it's not - I'll give it a shot